### PR TITLE
Shared volume support

### PR DESCRIPTION
**Pull request by kotfic**
Still need to implement:

Features:
+ [x] detaching a volume
+ [x] deleting a volume
+ [x] detaching all volumes when a cluster is terminated 
+ [x] mocked out tests for create/attach/detach/delete/status

Cleanup:
+ [x] refactor common cluster/volume functions into utils.py
+ [x] remove dynamic_inventory.py in favor of EC2Provider.py

### CODE REVIEW

File: girder/cumulus/server/volume.py [hunk 1]
```diff
@@ -63,8 +73,52 @@ def _create_ebs(self, body, zone):
...
+        return profile, secret
+
+
+    @access.user
+    def patch(self, id, params):
+        body = getBodyJson()
+        user = self.getCurrentUser()
+
+        volume = self._model.load(id, user=user, level=AccessType.WRITE)
+
```
Comment (cjh1) [inline girder/cumulus/server/volume.py]: We should probably update this to use the @loadmodel decorator.

File: girder/cumulus/server/volume.py [hunk 2]
```diff
@@ -63,8 +73,52 @@ def _create_ebs(self, body, zone):
         return self.model('volume', 'cumulus').create_ebs(user, profileId, name,
                                                           zone, size, fs)
 
+    # NOTE:  this is a blind copy from cluster_adapters.py
+    # should really be refactored into something more reusable
+    def _get_profile(self, profile_id):
```
Comment (cjh1) [inline girder/cumulus/server/volume.py]: We can add a module to the cumulus.common package

File: girder/cumulus/server/utility/cluster_adapters.py [hunk 3]
```diff
@@ -247,11 +248,6 @@ def start(self, request_body):
         provision_playbook_params['ansible_ssh_user'] = provision_ssh_user
         provision_playbook_params['cluster_state'] = 'running'
 
-        # If we are launching sge, then set the name of the master node
-        master_name = None
-        if launch_playbook == self.DEFAULT_PLAYBOOK:
-            master_name = 'head'
-
         cumulus.ansible.tasks.cluster.start_cluster \
```
Comment (cjh1) [inline girder/cumulus/server/utility/cluster_adapters.py]: Why remove this?

File: girder/cumulus/server/models/aws.py [hunk 4]
```diff
@@ -132,12 +133,14 @@ def validate(self, doc):
...
-                       region_name, availability_zone, public_ips):
+    def create_profile(self, userId, name, profile_type, access_key_id,
+                       secret_access_key, region_name, availability_zone,
+                       public_ips):
 
         user = getCurrentUser()
         profile = {
             'name': name,
+            'type': profile_type,
             'accessKeyId': access_key_id,
```
Comment (cjh1) [inline girder/cumulus/server/models/aws.py]: What is the type field for? Aren't these always 'ec2' profiles? If this is part of a move to make the profiles more generic ( with a view to other providers in the future ) I think we should look at some renaming ...

File: cumulus/ansible/tasks/playbooks/volumes/ec2/attach.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            device_name: /dev/xvdf
+          delegate_to: localhost
+          register: vol
+
+        - filesystem:
+            fstype: ext4
+            dev: /dev/xvdf
+          become: yes
+          become_user: root
+
```
Comment (cjh1) [inline cumulus/ansible/tasks/playbooks/volumes/ec2/attach.yml]: I assume the file system is only create is it doesn't already exist.

File: cumulus/ansible/tasks/providers/base.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    PENDING = "pending"
+    RUNNING = "running"
+    SHUTTINGDOWN = "shutting-down"
+    TERMINATED = "terminated"
+    STOPPING = "stopping"
+    STOPPED = "stopped"
+
+
+class Provider(object):
+    __provider_registry__ = {}
```
Comment (cjh1) [inline cumulus/ansible/tasks/providers/base.py]: Should this be ClusterProvider? InstanceProvider? CloudProvider?  Provider seems overly generic to me.

File: cumulus/ansible/tasks/utils.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        check_status(r)
+
+
+def check_ansible_return_code(returncode, cluster, girder_token):
+    if returncode != 0:
+        check_status(requests.patch('%s/clusters/%s' %
+                                    (cumulus.config.girder.baseUrl,
+                                     cluster['_id']),
+                                    headers={'Girder-Token': girder_token},
+                                    json={'status': 'error'}))
```
Comment (cjh1) [inline cumulus/ansible/tasks/utils.py]: Nice, I like this refactor, always thought there was too much in the cluster.py module.

File: cumulus/ansible/tasks/providers/ec2.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+            raise Exception('Required env var %s not given to Cumulus'
+                            'inventory.' % required_env_var)
+
+    p = Provider({
+        "accessKeyId": os.environ.get("AWS_ACCESS_KEY_ID"),
+        "secretAccessKey": os.environ.get("AWS_SECRET_ACCESS_KEY"),
+        "type": "ec2"
+    })
+
+    print(json.dumps(p.get_inventory(os.environ.get('CLUSTER_ID'))))
```
Comment (cjh1) [inline cumulus/ansible/tasks/providers/ec2.py]: When does this module get run on the command line?

File: girder/cumulus/server/utility/cluster_adapters.py (hunk 3 above)
Comment (cjh1) [inline girder/cumulus/server/utility/cluster_adapters.py]: I guess this information is now provided by the "Provider"

File: cumulus/ansible/tasks/volume.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                'AWS_SECRET_ACCESS_KEY': secret_key,
+                'ANSIBLE_HOST_KEY_CHECKING': 'false',
+                'ANSIBLE_CALLBACK_PLUGINS': get_callback_plugins_path(),
+                'ANSIBLE_LIBRARY': get_library_path()})
+
+    inventory = simple_inventory({'head': [instance['public_ip']]})
+
+    with inventory.to_tempfile() as inventory_path:
+        run_playbook(playbook, inventory_path,
+                     extra_vars, verbose=2, env=env)
```
Comment (cjh1) [inline cumulus/ansible/tasks/volume.py]: Where do the log messages end up for these ansible tasks. I can see that in the case of attaching a volume to a cluster we can write the log to the cluster log, but what about creating a volume? Should we add a volume log? I think its important that this information is captured somewhere so we can provide it to the user.

**Comment by cjh1** [comment]
@kotfic This is shaping up nicely. Seeing the playbook for attaching a volume and how simple its is to ensure we have a filesystem and mounting it validates our decision to use Ansible here in my opinion.

**Comment by patrickoleary** [comment]
This looks good!

File: cumulus/ansible/tasks/cluster.py [hunk 10]
```diff
@@ -137,10 +137,10 @@ def provision_cluster(playbook, cluster, profile, secret_key, extra_vars,
 @command.task
 def start_cluster(launch_playbook, provision_playbook, cluster, profile,
                   secret_key, launch_extra_vars, provision_extra_vars,
-                  girder_token, log_write_url, master_name=None):
+                  girder_token, log_write_url):
```
Comment (cjh1) [inline cumulus/ansible/tasks/cluster.py]: Note the removal of master_name will break our current taskflows, no big deal we just need to remember to update them when this branch lands.

**Comment by cjh1** [comment]
When testing I ran into the follow:

  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "cumulus/taskflow/__init__.py", line 117, in wrapped
    return func(celery_task, *args, **kwargs)
  File "/opt/hpccloud/hpccloud/server/taskflows/hpccloud/taskflow/utility/__init__.py", line 146, in setup_cluster
    cluster = create_ec2_cluster(task, cluster, profile, ami)
  File "/opt/hpccloud/hpccloud/server/taskflows/hpccloud/taskflow/utility/__init__.py", line 127, in create_ec2_cluster
    launch_params, provision_params, girder_token, log_write_url)
  File "/usr/local/lib/python2.7/dist-packages/celery/local.py", line 188, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 439, in __protected_call__
    return orig(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 420, in __call__
    return self.run(*args, **kwargs)
  File "cumulus/ansible/tasks/cluster.py", line 143, in start_cluster
    launch_extra_vars, girder_token, log_write_url, 'launched')
  File "/usr/local/lib/python2.7/dist-packages/celery/local.py", line 188, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 439, in __protected_call__
    return orig(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 420, in __call__
    return self.run(*args, **kwargs)
  File "cumulus/ansible/tasks/cluster.py", line 182, in launch_cluster
    'host': master.public_ip
AttributeError: 'dict' object has no attribute 'public_ip'

Looks like for some reason the public_ip is not being populated in the inventory.

**Comment by cjh1** [comment]
I think I know the reason for the above error, I was creating a cluster with a single node ( which will only work when #224 is merged )

**Comment by cjh1** [comment]
Scratch that theory, still fails with my fix or with a two node cluster.

**Comment by cjh1** [comment]
I am getting further now, but still run into a issue:

```
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "cumulus/taskflow/__init__.py", line 117, in wrapped
    return func(celery_task, *args, **kwargs)
  File "/opt/hpccloud/hpccloud/server/taskflows/hpccloud/taskflow/utility/__init__.py", line 147, in setup_cluster
    cluster = create_ec2_cluster(task, cluster, profile, ami)
  File "/opt/hpccloud/hpccloud/server/taskflows/hpccloud/taskflow/utility/__init__.py", line 128, in create_ec2_cluster
    launch_params, provision_params, girder_token, log_write_url)
  File "/usr/local/lib/python2.7/dist-packages/celery/local.py", line 188, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 439, in __protected_call__
    return orig(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 420, in __call__
    return self.run(*args, **kwargs)
  File "cumulus/ansible/tasks/cluster.py", line 151, in start_cluster
    'provisioned')
  File "/usr/local/lib/python2.7/dist-packages/celery/local.py", line 188, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 439, in __protected_call__
    return orig(self, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/celery/app/task.py", line 420, in __call__
    return self.run(*args, **kwargs)
  File "cumulus/ansible/tasks/cluster.py", line 118, in provision_cluster
    playbook_variables = get_playbook_variables(cluster, profile, extra_vars)
  File "cumulus/ansible/tasks/cluster.py", line 70, in get_playbook_variables
    'ansible_ssh_private_key': _key_path(profile)
  File "cumulus/ssh/tasks/key.py", line 34, in _key_path
    return os.path.join(cumulus.config.ssh.keyStore, str(profile['_id']))
KeyError: '_id'
```

If I look at 'profile' it is indeed missing the '_id' field!

File: girder/cumulus/server/models/aws.py (hunk 4 above)
Comment (kotfic) [inline girder/cumulus/server/models/aws.py]: This is a look towards more generic profiles.  Currently ```type``` is what determines the Provider class we load. What did you have in mind for naming?

File: cumulus/ansible/tasks/playbooks/volumes/ec2/attach.yml (hunk 5 above)
Comment (kotfic) [inline cumulus/ansible/tasks/playbooks/volumes/ec2/attach.yml]: That is correct.  It is only created if it doesn't already exist.

File: cumulus/ansible/tasks/providers/base.py (hunk 6 above)
Comment (kotfic) [inline cumulus/ansible/tasks/providers/base.py]: CloudProvider sounds fine to me. The purpose of the class(es) is to provide a consistent read-only interface to cloud providers like AWS. This lets us check state,  fail early and build/modify ansible inventories

File: cumulus/ansible/tasks/providers/ec2.py (hunk 8 above)
Comment (kotfic) [inline cumulus/ansible/tasks/providers/ec2.py]: This is part of a plan to do away with dynamic_inventory/ec2.py and use providers/ec2.py for all things related to pulling state from AWS. dynamic_inventory/ec2.py was originally designed to be run like ```ansible-playbook -i ec2.py foo.yml```  this ensures that providers/ec2.py replicates that functionality

File: cumulus/ansible/tasks/volume.py (hunk 9 above)
Comment (kotfic) [inline cumulus/ansible/tasks/volume.py]: because girder_log_write_url isn't set log information just ends up in the celery stdout/stderr.  If we set up an endpoint and pass it in,  the callback_module will post info to that location. We just need to decide where that will be.

File: cumulus/constants.py [hunk 11]
```diff
@@ -44,8 +44,13 @@ def is_valid_type(type):
 
 
 class VolumeState:
+    PROVISIONING = 'provisioning'
     AVAILABLE = 'available'
+    ATTACHING = 'attaching'
+    DETACHING = 'detaching'
     INUSE = 'in-use'
+    TERMINATING = 'terminating'
```
Comment (cjh1) [inline cumulus/constants.py]: terminating?

File: tests/integration/ec2_volume_integration_test.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+        super(EC2VolumeIntegrationTest, self).__init__(
+            name, girder_url, girder_user, girder_password, aws_access_key_id,
+            aws_secret_access_key)
+        self._volume_id = None
+        self._profile_id = "576acbf30640fd0a56df35e1"
+        self._cluster_id = "576acbf70640fd0a56df35e3"
+
+    def tearDown(self):
+        super(EC2VolumeIntegrationTest, self).tearDown()
+
```
Comment (cjh1) [inline tests/integration/ec2_volume_integration_test.py]: Should probably ensure that the volume is cleaned up.

File: cumulus/ansible/tasks/cluster.py [hunk 13]
```diff
@@ -122,25 +32,25 @@ def provision_cluster(playbook, cluster, profile, secret_key, extra_vars,
...
+                'ANSIBLE_CALLBACK_PLUGINS': get_callback_plugins_path()})
 
     inventory = os.path.join(os.path.dirname(__file__), 'dynamic_inventory')
 
     ansible = run_playbook(playbook, inventory, playbook_variables,
                            env=env, verbose=3)
 
-    check_ansible_return_code(ansible, cluster, girder_token)
+    # todo cluster statuses should be updated in celery tasks?
     check_girder_cluster_status(cluster, girder_token, post_status)
```
Comment (cjh1) [inline cumulus/ansible/tasks/cluster.py]: Is this todo for another PR?

File: cumulus/ansible/tasks/utils.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+    while True:
+        reads = [p.stdout.fileno(), p.stderr.fileno()]
+        ret = select.select(reads, [], [])
+
+        for fd in ret[0]:
+            if fd == p.stdout.fileno():
+                logger.info(p.stdout.readline())
+            if fd == p.stderr.fileno():
+                logger.error(p.stderr.readline())
```
Comment (cjh1) [inline cumulus/ansible/tasks/utils.py]: Can we filter out the empty stderr line we are seeing?

File: cumulus/ansible/tasks/volume.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+    playbook = os.path.join(get_playbook_directory(),
+                            'volumes', 'ec2', 'create.yml')
+
+    extra_vars = {
+        'girder_volume_id': volume['_id'],
+        'profile_id': profile['_id'],
+        'volume_name': volume['name'],
+        'volume_size': volume['size'],
+        'volume_zone': volume['zone'],
```
Comment (cjh1) [inline cumulus/ansible/tasks/volume.py]: Doesn't the zone have to match our profile? i.e. if our access key is in us-east-2 we can't create a volume in us-west-2?

File: cumulus/ansible/tasks/volume.py (hunk 9 above)
Comment (cjh1) [inline cumulus/ansible/tasks/volume.py]: So we need to resolve this. Probably need to a volume log? Thoughts?

File: cumulus/aws/ec2/__init__.py [hunk 16]
```diff
@@ -7,13 +7,4 @@ class ClientErrorCode:
...
-    aws_secret_access_key = profile['secretAccessKey']
-    region_name = profile.get('regionName')
-
-    client = boto3.client(
-        'ec2', aws_access_key_id=aws_access_key_id,
-        aws_secret_access_key=aws_secret_access_key,
-        region_name=region_name)
-
-    return client
+    return CloudProvider(profile)
```
Comment (cjh1) [inline cumulus/aws/ec2/__init__.py]: Nice!

File: girder/cumulus/server/models/aws.py (hunk 4 above)
Comment (cjh1) [inline girder/cumulus/server/models/aws.py]: CloudProfile?

File: cumulus/ansible/tasks/volume.py [hunk 17]
```diff
@@ -0,0 +1,137 @@
+from cumulus.celery import command
```
Comment (cjh1) [inline cumulus/ansible/tasks/volume.py]: Can we add a copyright header?

File: cumulus/ansible/tasks/utils.py [hunk 18]
```diff
@@ -0,0 +1,111 @@
+import pkg_resources as pr
```
Comment (cjh1) [inline cumulus/ansible/tasks/utils.py]: Header?

File: cumulus/ansible/tasks/providers/ec2.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
+from cumulus.constants import VolumeState
```
Comment (cjh1) [inline cumulus/ansible/tasks/providers/ec2.py]: Header?

File: cumulus/ansible/tasks/providers/__init__.py [hunk 20]
```diff
@@ -0,0 +1,4 @@
+from base import CloudProvider, InstanceState
```
Comment (cjh1) [inline cumulus/ansible/tasks/providers/__init__.py]: Header?

File: tests/integration/ec2_volume_integration_test.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+###############################################################################
+#  Copyright 2015 Kitware Inc.
```
Comment (cjh1) [inline tests/integration/ec2_volume_integration_test.py]: 2015=>2016

File: tests/integration/ec2_volume_integration_test.py [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+        super(EC2VolumeIntegrationTest, self).__init__(
+            name, girder_url, girder_user, girder_password, aws_access_key_id,
+            aws_secret_access_key)
+        self._volume_id = None
+        self._profile_id = None
+        self._cluster_id = None
+
+    def tearDown(self):
+        super(EC2VolumeIntegrationTest, self).tearDown()
+
```
Comment (cjh1) [inline tests/integration/ec2_volume_integration_test.py]: Should delete volume ( ensure if was deleted ...)

File: cumulus/ansible/tasks/cluster.py [hunk 23]
```diff
@@ -122,25 +32,25 @@ def provision_cluster(playbook, cluster, profile, secret_key, extra_vars,
...
-                  girder_token, log_write_url, master_name=None):
+                  girder_token, log_write_url):
 
-    run_ansible(launch_playbook, cluster, profile, secret_key,
-                launch_extra_vars, girder_token, log_write_url, 'launched')
+    launch_cluster(launch_playbook, cluster, profile, secret_key,
+                   launch_extra_vars, girder_token, log_write_url, 'launched')
 
     # todo cluster statuses should be updated in celery tasks?
     check_girder_cluster_status(cluster, girder_token, 'provisioning')
```
Comment (cjh1) [inline cumulus/ansible/tasks/cluster.py]: todo for next PR?

File: cumulus/ansible/tasks/providers/ec2.py [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+            .find(aws_volume.describe_status())[0].value
+
+        if status != 'ok':
+            if status == 'insufficient-data':
+                volume['state'] = VolumeState.PROVISIONING
+            else:
+                volume['state'] = VolumeState.ERROR
+
+        # TODO add 'instance_id' here (None if not attached)
+        if aws_volume.attachments:
```
Comment (cjh1) [inline cumulus/ansible/tasks/providers/ec2.py]: TODO?

File: girder/cumulus/server/utility/cluster_adapters.py [hunk 25]
```diff
@@ -118,26 +117,6 @@ def validate(self):
...
-        if profile is None:
-            raise ValidationException('Profile must be specified!')
-
-        profile['_id'] = str(profile['_id'])
-
-        return profile, secret
-
     def launch(self):
         # if id is None // Exception
         # TODO: add assert if status > launching, error
```
Comment (cjh1) [inline girder/cumulus/server/utility/cluster_adapters.py]: TODO?

File: girder/cumulus/server/volume.py [hunk 26]
```diff
@@ -77,66 +112,42 @@ def create(self, params):
...
-                                                    user=getCurrentUser())
+        profile, secret_key = _get_profile(profile_id)
+
         if not profile:
             raise RestException('Invalid profile', 400)
 
-        client = get_ec2_client(profile)
-
+        # TODO:  Validate zone using provider
         if 'zone' in body:
```
Comment (cjh1) [inline girder/cumulus/server/volume.py]: Looks like this TODO can be removed?

File: girder/cumulus/server/models/aws.py (hunk 4 above)
Comment (kotfic) [inline girder/cumulus/server/models/aws.py]: how about cloud_provider?

File: cumulus/ansible/tasks/volume.py (hunk 9 above)
Comment (kotfic) [inline cumulus/ansible/tasks/volume.py]: I've created an issue to track this https://github.com/Kitware/cumulus/issues/237 lets put it into a separate PR though so we can shut down this monster

File: girder/cumulus/server/models/aws.py (hunk 4 above)
Comment (cjh1) [inline girder/cumulus/server/models/aws.py]: cloud-provider? ( I prefer kebab case for endpoints )

File: cumulus/ansible/tasks/volume.py (hunk 9 above)
Comment (cjh1) [inline cumulus/ansible/tasks/volume.py]: Sounds good

File: cumulus/ansible/tasks/volume.py (hunk 15 above)
Comment (kotfic) [inline cumulus/ansible/tasks/volume.py]: Access keys provide access to all zones,  its true you can't attach a volume to a node in a different zone,  but if you wanted to create a volume first (before the cluster)  or wanted to add an existing volume to girder then you need to be able to specify a volume specific zone.

**Comment by cjh1** [comment]
@kotfic PTAL at my commits

**Comment by kotfic** [comment]
:clap:
